### ISSUE TITLE
Rebase not yet merged PR's

### ISSUE THREAD

**Issue by deivid-rodriguez** [gocardless/bump#213]
Hi! Thanks for bump. It works! :)

I have a feature request. I would like an option to tell "bump" to rebase previously opened PR's that have not been merged yet, and that are no longer up to date with master. Otherwise, we can't really be sure that the dependency update doesn't break our app.

Regards!

**Comment by greysteil** [comment]
Thanks for the kind words @deivid-rodriguez.

Bump doesn't have a database, so doing those rebases correctly is tricky - it needs to keep track of whether you've updated the PR yourself, so it doesn't overwrite your changes. We wanted to keep it lightweight and free to self-host on Heroku, so built [Dependabot](https://dependabot.com) as a more fully-featured version with all the extra gubbins, including automatic rebasing. It's totally free for everyone at the moment, but at some point we need to start charging for private repos (intended price list on the website). We'll also probably open-source the code to that at some point, too, which would make self-hosting it possible, although the infrastructure to host it costs about the same as a private org would pay.

Hope that helps!

**Comment by deivid-rodriguez** [comment]
Yes, I didn't know dependabot had that feature. I will try it. Thanks! :)

**Comment by deivid-rodriguez** [comment]
@greysteil I tried dependabot. It works grate, but it does not seem to do the rebasing I was expecting. When it detects merge conflicts, it _does_ close the PR and opens a new updated one. But I'd like it to do that every time it detects an outdated branch, even if it merges cleanly. I guess that's not possible at the moment?

**Comment by greysteil** [comment]
Is this for working with protected branches where only fully up-to-date PRs can be merged? We've had that request from someone else and it makes sense to me. Slightly tricky on our side, as it means we could be doing a LOT of PR rebasing is the repo is very active...!

**Comment by deivid-rodriguez** [comment]
The rationale is that unless the PR is up to date with master, it has not been tested against it, so it's less safe to merge. I think it's the same rationale for the "only allow merging up-to-date PRs on protected branches" github feature.

To my understanding, you're not really messing with rebasing PR's anyways. You just close and open a fresh one, right?

**Comment by greysteil** [comment]
Hmmm, that makes sense. Perhaps having a "rebase against master now" button in the PR message would work, and save us from doing rebases that aren't necessary?

RE: rebasing vs/closing - we only close if there's a new (superseding) version; for all other cases we do a rebase. During the rebase we put a note at the bottom of the PR message, saying it's currently being rebased, and then remove that note once the rebase is complete. Easiest way to see it is to push a change to your Gemfile / Gemfile.lock that will conflict with an open Dependabot PR.

**Comment by deivid-rodriguez** [comment]
> mm, that makes sense. Perhaps having a "rebase against master now" button in the PR message would work, and save us from doing rebases that aren't necessary?

That would definitely work for me! :+1: 

> RE: rebasing vs/closing - we only close if there's a new (superseding) version; for all other cases we do a rebase. During the rebase we put a note at the bottom of the PR message, saying it's currently being rebased, and then remove that note once the rebase is complete. Easiest way to see it is to push a change to your Gemfile / Gemfile.lock that will conflict with an open Dependabot PR.

I see, thanks for explaining.

**Comment by greysteil** [comment]
Quick update on this: I've been digging through the GitHub docs on their [Git Data API](https://developer.github.com/v3/git/) and it's painful - there's no simple way to do a rebase, particularly when a file has changed but has no conflicts.

Have you looked at using a bot like Bundler's [Homu](https://github.com/bundler/homu) to achieve the same ends with your CI server, rather than with rebases?

**Comment by deivid-rodriguez** [comment]
Yep, I've been wanting to experiment with something like Homu for a while, but this seemed simpler for now, yes :)

What about a similar option that does not involve rebases? Like a "reopen on top of current master" or something like that.

Note that this is not a big deal anyways since Github provides this "update branch" button that does something very similar. I just don't like that it creates a merge commit.

**Comment by gotjosh** [comment]
@deivid-rodriguez @greysteil We have now deprecated bump in favour of [Dependabot](dependabot.com) I'll be closing this issue as I believe its no longer relevant.

**Comment by greysteil** [comment]
(But I'll keep it in mind for Dependabot!)

**Comment by greysteil** [comment]
@deivid-rodriguez - I just added this to Dependabot. If you comment `@dependabot rebase?` it'll trigger a rebase for you. Not documented yet as I want to be 100% sure that's the best interface - let me know if you have any feedback!
